tokenizer/scanner.py

```diff
@@ -1,3 +1,78 @@
 
-def scan(code):
-    print("scanner")
+def scan(filename):
+    print("scanner start")
+    token_specification = [
+        ('KEYWORD', ['if', 'then', 'else', 'while', 'do', 'end', 'procedure', 'var', 'begin', 'print']),
+        ('OPERATOR', ['=','+', '-', '*', '/', '==', '!=', '<=', '>=', '<', '>', 'and', 'or', 'not']),
+        ('SEPARATOR', [';']),
+        ('LPAR', ['(']),
+        ('RPAR', [')'])
+    ]
+    
+    tokens = []
+    
+    with open(filename, 'r', encoding='utf-8') as file:
+        for line in file:
+            input_string = line.strip()
+            i = 0
+            while i < len(input_string):
+                # ignore whitespace
+                if input_string[i].isspace():
+                    i += 1
+                    continue
+
+                matched = False
+
+                # Keyword
+                for token_type, values in token_specification:
+                    if token_type == 'KEYWORD':
+                        for keyword in values:
+                            if input_string[i:i+len(keyword)] == keyword and (i+len(keyword) == len(input_string) or not input_string[i+len(keyword)].isalnum()):
+                                tokens.append((token_type, keyword))
+                                i += len(keyword)
+                                matched = True
+                                break
+                        if matched:
+                            break
+
+                # Operator
+                if not matched:
+                    for operator in token_specification[1][1]:
+                        if input_string[i:i+len(operator)] == operator:
```

Comment:
If operator is "==", "=" will be matches first, and then break out of the loop

Reply:
This is indeed a problem, I will fix it later.
(cant check ==)